### ISSUE TITLE
Unable to configure extension when using a template deployment

### ISSUE THREAD

**Issue by stevenewey** [Azure/azure-docker-extension#44]
I suspect I've got the JSON format wrong for configuring the extension, but I've not been able to find any specific documentation for what I'm trying to do. I'm bringing up a number of VMs and trying to configure Docker on them using a resource-group based templated deployment.

The relevant part of my template is:
```
{
    "apiVersion": "2015-05-01-preview",
    "dependsOn": [
        "Microsoft.Compute/virtualMachines/VMNAME"
    ],
    "location": "[resourceGroup().location]",
    "name": "VMNAME/dockerExtension",
    "properties": {
        "autoUpgradeMinorVersion": true,
        "publisher": "Microsoft.Azure.Extensions",
        "settings": {
            "certs": {
                "ca": "...",
                "cert": "...",
                "key": "..."
            },
            "login": {
                "email": "...",
                "password": "...",
                "username": "..."
            },
            "docker": {
                "options": [
                    "-D",
                    "--dns=8.8.8.8",
                    "--storage-driver=aufs",
                    "--graph=/mnt"
                ],
                "port": "2376"
            }
        },
        "type": "DockerExtension",
        "typeHandlerVersion": "1.0"
    },
    "type": "Microsoft.Compute/virtualMachines/extensions"
}
```
And I'm receiving this error in /var/log/azure/Microsoft.Azure.Extensions.DockerExtension/1.0.1507151643/docker-extension.log

```
[DockerExtension] 2015/07/21 14:14:15 ----------------------------------------
[DockerExtension] 2015/07/21 14:14:15 Extension handler launch args: "./bin/docker-extension enable"
[DockerExtension] 2015/07/21 14:14:15 distro info: Ubuntu 14.04
[DockerExtension] 2015/07/21 14:14:15 using distro driver: driver.UbuntuUpstartDriver
[DockerExtension] 2015/07/21 14:14:15 user: root uid:0 gid:0
[DockerExtension] 2015/07/21 14:14:15 env['PATH'] = /usr/local/sbin:/usr/local/bin:/usr/bin:/usr/sbin:/sbin:/bin
[DockerExtension] 2015/07/21 14:14:15 + starting: 'enable'
[DockerExtension] 2015/07/21 14:14:15 ERROR: error parsing handler settings: error deserializing public settings for handler: json: cannot unmarshal number into Go value of type string
```

Any hints as to what I'm doing wrong, or even better, an addition to the README to describe using the extension from a templated deployment would be gratefully received.

Thanks
Stephen

**Comment by ahmetalpbalkan** [comment]
Hi Stephen,

The error looks obscure. Can you please add your `/var/lib/waagent/[extension-name]/config/0.settings` file (by redacting your certs/password) as well.

Meanwhile, you should move  `certs`  and `login` objects into `"protectedSettings"` key in your template (not `"settings"`) as they are secrets. Otherwise they will be discarded. This is a nice reminder we should probably add a full fledged ARM template example as well.

**Comment by stevenewey** [comment]
Thanks @ahmetalpbalkan. I've now moved my certs and login as you suggest. The contents of 0.settings are as follows...

```
{
  "runtimeSettings": [
    {
      "handlerSettings": {
        "publicSettings": {"docker":{"certs":{"ca":"...","cert":"...","key":"..."},"login":{"email":"...","password":"...","username":"..."},"options":["-D","--dns=8.8.8.8","--storage-driver=aufs","--graph=/mnt"],"port":2376}}
      }
    }
  ]
}
```

Just a guess, but I'm assuming the port number being an integer rather than a string is the problem? It's a string in my deployment template :)

**Comment by ahmetalpbalkan** [comment]
@stevenewey Hmm, port number is my #1 suspicion at this point. It should be string at all times. So was it a `string` your template and you found it converted to `int` here somehow or was it string all along?

**Comment by stevenewey** [comment]
It was a string in my deployment template. I've actually just tried removing port altogether from the template and re-running the deployment. And I actually get the same error and see the same data in 0.settings (and the file was definitely updated to a newer modified time).

**Comment by ahmetalpbalkan** [comment]
Hmm what do you mean you see the same data in 0.settings, was it unchanged although you updates the template?

Are you reusing the VM between deployments of te DockerExtension? If so I believe you need to first uninstall the extension and add it to the VM again. Is it possible to start from a clean VM for you in this case?

**Comment by stevenewey** [comment]
@ahmetalpbalkan The data was unchanged, though the modified time on the file on disk had updated. So I tried your advice and just deleted the VM and things magically started working when I reran the deploy! I then deleted and re-created again, this time with the port back in (as a string) and it also worked, so the issue was I'd used a integer at some previous attempt.

Sorry about that! And thanks for you help.
